### DISCUSSION TITLE
Set Multicast IPv6 interface

### DISCUSSION THREAD

**Discussion by tropxy** [tokio-rs/tokio#4603] [General]
Hi,

I am setting up an IPv6 Multicast server and client and I noticed that neither [tokio](https://docs.rs/tokio/latest/tokio/index.html)::[net](https://docs.rs/tokio/latest/tokio/net/index.html)::[UdpSocket](https://docs.rs/tokio/latest/tokio/net/struct.UdpSocket.html#) neither the [std](https://doc.rust-lang.org/stable/std/index.html)::[net](https://doc.rust-lang.org/stable/std/net/index.html)::[UdpSocket](https://doc.rust-lang.org/stable/std/net/struct.UdpSocket.html#) provide a way to set the multicast interface as it is done in the socket2 crate: https://docs.rs/socket2/latest/socket2/struct.Socket.html#method.set_multicast_if_v6

As far as I know, the server is required to join the multicast group, whilst the client doesnt. The client needs just to specify the interface that wants to use to send the multicast packets from and when calling the `send_to` function, specify the remote address which shall be the concatenation of the multicast group address and the port.

However, due to the underlying UdpSocket API of tokio, we are forced to use the `join_multicast_v6` method. Is this correct?
Shouldnt we have a way to set the multicast if the client wants to use without joining the interface to the multicast group?

An example of it should look like using socket2 crate is provided here: https://github.com/bluejekyll/multicast-example/blob/master/src/lib.rs#L161

**Comment by Darksonn** [comment]
You can use the socket2 methods on a Tokio socket.

**Comment by tropxy** [comment]
Hi,
yup, that is what I also found out, thanks. But shouldnt tokio expose the calls in the API?

EDIT: for future reference, here is an example of how to do it: https://github.com/henninglive/tokio-udp-multicast-chat/blob/master/src/main.rs#L29

beware that the socket2 version used is an old one and `into_udp_socket` was deprecated in favour of `std::net::UdpSocket.from()` trait

**Reply by Darksonn** [reply]
There are lots of socket APIs that we don't expose. You are welcome to open an issue asking us to add this particular API.

**Reply by tropxy** [reply]
Ok thanks for that @Darksonn. I will do so ;)
